### PR TITLE
Fix string index with negative numbers

### PR DESCRIPTION
**Pull request by tpoterba**
Should speed it up, too!

cc @danking

### CODE REVIEW

**Comment by danking** [comment]
@tpoterba ah, python supports this nonsense:
```
>>> "foo"[10000:100002]
''
```

**Comment by danking** [comment]
And `ir.StringSlice` does not

File: hail/src/test/scala/is/hail/expr/ir/StringFunctionsSuite.scala [hunk 1]
```diff
@@ -67,24 +67,74 @@ class StringFunctionsSuite extends TestNGSuite {
...
-      Array(NA(TStruct("x" -> TInt32())), TStruct("x" -> TInt32())),
-      Array(F32(3.14f), TFloat32()),
-      Array(I64(7), TInt64()),
-      Array(IRArray(1, null, 5), TArray(TInt32())),
-      Array(MakeTuple(Seq(1, NA(TInt32()), 5.7)), TTuple(TInt32(), TInt32(), TFloat64()))
-    )
-
-    @Test(dataProvider = "str")
-    def str(annotation: IR, typ: Type) {
-      assertEvalsTo(invoke("str", annotation), {val a = eval(annotation); if (a == null) null else typ.str(a) })
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/test/scala/is/hail/expr/ir/StringFunctionsSuite.scala]: I think it wasn't enabled

[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/test/scala/is/hail/expr/ir/StringFunctionsSuite.scala]: because of that

**Comment by tpoterba** [comment]
actually, is this a problem in our interpreter?

**Comment by tpoterba** [comment]
should we relax this check?

**Comment by danking** [comment]
Which check? FWIW, [array slicing](https://github.com/hail-is/hail/blob/2edc6d2ce2c20fda56f7e032d4b1890c186f7d52/hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala#L287-L302) goes through the same rigamarole to maintain python compatibility.

**Comment by tpoterba** [comment]
I think we need to relax the bounds checks on stringslice (and array slice)

**Comment by tpoterba** [comment]
```
In [2]: "A"[-100:-90]
Out[2]: ''

In [3]: "A"[10:8]
Out[3]: ''

In [4]: "ABC"[3:2]
Out[4]: ''
```

**Comment by tpoterba** [comment]
it seems literally impossible for string slicing to fail in python

**Comment by tpoterba** [comment]
ah, crap. I'm hitting the interpret issue about evaluation both sides of `&`.

**Comment by tpoterba** [comment]
The bug that this PR fixes meant that `_is_snp_transition` succeeded even when it shouldn't:

```
def _is_snp_transition(ref, alt) -> BooleanExpression:
    indices = hl.range(0, ref.length())
    return hl.any(lambda i: ((ref[i] != alt[i]) & (((ref[i] == 'A') & (alt[i] == 'G')) |
                                                   ((ref[i] == 'G') & (alt[i] == 'A')) |
                                                   ((ref[i] == 'C') & (alt[i] == 'T')) |
                                                   ((ref[i] == 'T') & (alt[i] == 'C')))), indices)
```

(a precondition is that the lengths are the same)

File: hail/src/main/scala/is/hail/expr/ir/Interpret.scala [hunk 2]
```diff
@@ -653,6 +655,30 @@ object Interpret {
       case Die(message, typ) => fatal(message)
       case ir@ApplyIR(function, functionArgs, conversion) =>
         interpret(ir.explicitNode, env, args, agg)
+      case ApplySpecial("||", Seq(left_, right_)) =>
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/Interpret.scala]: cc @catoverdrive - I had to add these rules to make python tests pass after fixing this bug.

Comment (catoverdrive) [reply hail/src/main/scala/is/hail/expr/ir/Interpret.scala]: ah, good call. That should let us fix the places where we want to check this in other tests, too.

**Pull request merged**